File: src/main/java/leetcode/problems/medium/House_Robber_III_337.java
package leetcode.problems.medium;

import leetcode.problems.util.TreeNode;

import java.util.HashMap;
import java.util.Map;

/**
 * 337. House Robber III
 * https://leetcode.com/problems/house-robber-iii/
 *
 * The thief realized that all houses in this place form a binary tree.
 * It will automatically contact the police if two directly-linked houses were broken into on the same night.
 * Given the root of the binary tree, return the maximum amount of money the thief can rob without alerting the police.
 * Constraints:
 *      The number of nodes in the tree is in the range [1, 10^4].
 *      0 <= Node.val <= 10^4
 */
public class House_Robber_III_337 {

    Map<TreeNode, Integer> cache = new HashMap<>();

    /**
     * Top-down approach:
     * 2 decisions:
     *  either rob this one, and don't rob its children
     *  or don't rob this one
     *
     * Runtime: 3 ms, faster than 22.28%
     * Memory Usage: 38.7 MB, less than 62.35%
     */
    public int rob(TreeNode node) {
        if (node == null) {
            return 0;
        }
        // we decide to rob this one, or not

        int robThis = node.val;
        if (node.left != null) {
            if (node.left.left != null && !cache.containsKey(node.left.left)) {
                cache.put(node.left.left, rob(node.left.left));
            }
            if (node.left.right != null && !cache.containsKey(node.left.right)) {
                cache.put(node.left.right, rob(node.left.right));
            }
            robThis += cache.getOrDefault(node.left.left, 0);
            robThis += cache.getOrDefault(node.left.right, 0);
        }
        if (node.right != null) {
            if (node.right.left != null && !cache.containsKey(node.right.left)) {
                cache.put(node.right.left, rob(node.right.left));
            }
            if (node.right.right != null && !cache.containsKey(node.right.right)) {
                cache.put(node.right.right, rob(node.right.right));
            }
            robThis += cache.getOrDefault(node.right.left, 0);
            robThis += cache.getOrDefault(node.right.right, 0);
        }

        int robChildren = 0;
        if (node.left != null && !cache.containsKey(node.left)) {
            cache.put(node.left, rob(node.left));
        }
        robChildren += cache.getOrDefault(node.left, 0);
        if (node.right != null && !cache.containsKey(node.right)) {
            cache.put(node.right, rob(node.right));
        }
        robChildren += cache.getOrDefault(node.right, 0);

        int res = Math.max(robThis, robChildren);
        cache.put(node, res);
        return res;
    }

    // Top-down approach - Slow

    public int rob0(TreeNode root) {
        return max(root, false);
    }
    /**
     * 2 decisions:
     *  either rob this one, and don't rob its children
     *  or don't rob this one
     */
    int max(TreeNode node, boolean isParentRobbed) {
        if (node == null) {
            return 0;
        }
        if (isParentRobbed) {
            // cannot rob this
            int leftMax = cache.getOrDefault(node.left, max(node.left, false));
            int rightMax = cache.getOrDefault(node.right, max(node.right, false));
            return leftMax + rightMax;
        }
        // at this point, parent is not robbed, so we decide to rob this one, or not
        int robThis = node.val + max(node.left, true) + max(node.right, true);
        int leftMax = cache.getOrDefault(node.left, max(node.left, false));
        int rightMax = cache.getOrDefault(node.right, max(node.right, false));
        int max = Math.max(robThis, leftMax + rightMax);
        cache.put(node, max);
        return max;
    }

    public static void main(String[] args) {
        House_Robber_III_337 p = new House_Robber_III_337();
        System.out.println(p.rob(new TreeNode(3, new TreeNode(2, null, new TreeNode(3)), new TreeNode(3, null, new TreeNode(1)))));
    }
}

File: src/main/scala/leetcode/medium/House_Robber_iii_337.scala
package leetcode.medium

import scala.collection.mutable

case class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}

object House_Robber_iii_337 extends App {
  val cache = mutable.Map.empty[TreeNode, Int]
  def rob(root: TreeNode): Int = {
    if (root == null) {
      0
    } else if (cache.contains(root)) {
      cache(root)
    } else {
      var ll = 0
      var lr = 0
      var rr = 0
      var rl = 0
      if (root.left != null) {
        if (!cache.contains(root.left)) {
          cache.put(root.left, rob(root.left))
        }
        if (root.left.left != null) {
          if (!cache.contains(root.left.left)) {
            cache.put(root.left.left, rob(root.left.left))
          }
          ll = cache(root.left.left)
        }
        if (root.left.right != null) {
          if (!cache.contains(root.left.right)) {
            cache.put(root.left.right, rob(root.left.right))
          }
          lr = cache(root.left.right)
        }
      }
      if (root.right != null) {
        if (!cache.contains(root.right)) {
          cache.put(root.right, rob(root.right))
        }
        if (root.right.left != null) {
          if (!cache.contains(root.right.left)) {
            cache.put(root.right.left, rob(root.right.left))
          }
          rl = cache(root.right.left)
        }
        if (root.right.right != null) {
          if (!cache.contains(root.right.right)) {
            cache.put(root.right.right, rob(root.right.right))
          }
          rr = cache(root.right.right)
        }
      }
      val result = Math.max(
        cache.getOrElse(root.left, 0) + cache.getOrElse(root.right, 0),
        root.value + ll + lr + rr + rl
      )
      cache.put(root, result)
      result
    }
  }

  println(rob(TreeNode(3, TreeNode(2, null, TreeNode(3)), TreeNode(3, null, TreeNode(1)))))
}
